### PR TITLE
refactor: Sub-class `Session` from `db.Entity` to behave `dict`-compliant

### LINKED ISSUE #1147: Subclass `Session` from `db.Entity` to inherit all dict-features
> We generally should think about subclassing `Session` from `db.Entity`, which is a subclass of dict, so that the session can be used like an ordinary dict in all situations.

_Originally posted by @phorward in https://github.com/viur-framework/viur-core/pull/1140#pullrequestreview-2027977727_

### PR DESCRIPTION
**Pull request by KadirBalku**
Resolves #1147, replaces #1140.

### CODE REVIEW

**Review by phorward — CHANGES_REQUESTED** [review]
Hello @KadirBalku, thanks for your first pull request!

I've added some improvement suggestions.
Can you test your branch with other projects than "viur-trial" please, just for confidence.

**Review by phorward — CHANGES_REQUESTED** [review]
Hello @KadirBalku,

have you tested it with other projects?

And can you please take a look at the functions I've commented. Maybe they must be kept, or the internally use `__setitem__`. In latter case, all is fine, and the `changed`-flag is being set properly. 

Take a look into the [Python docs](https://docs.python.org/3/library/stdtypes.html#typesmapping) to get an overview about all dict-related functions.

**Review by phorward — CHANGES_REQUESTED** [review]
@KadirBalku looks good to me!
Please fix the unused import issue, then we can merge.

File: src/viur/core/session.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import datetime
-import hmac
+import hmac  # is this necessary ?
```
[the remark this answers is not in the document]
Comment (KadirBalku) [reply src/viur/core/session.py:2]: the comment is added by me, so i can remove it without hesitation ?

File: src/viur/core/session.py [hunk 2]
```diff
@@ -73,7 +72,11 @@ def load(self, req):
                     return False
 
                 self.cookie_key = cookie_key
-                self.session = data["data"]
+
+                # Here we have to use the pure db.Entity function to avoid changing the 'changed' flag
+                db.Entity.clear(self)
+                db.Entity.update(self, data["data"])
```
Comment (sveneberth) [inline src/viur/core/session.py:78]: I think a `super()` call would be more pythonic?
```suggestion
                super().clear()
                super().update(data["data"])
```

File: src/viur/core/session.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sveneberth) [reply src/viur/core/session.py:2]: yes, this is unnecessary and can be removed.

**Review by sveneberth — CHANGES_REQUESTED** [review]
What about `Session.pop()` and where is `Session.setdefault()` from https://github.com/viur-framework/viur-core/pull/1140?

File: src/viur/core/session.py [hunk 3]
```diff
@@ -164,6 +166,29 @@ def reset(self) -> None:
...
+        super().__ior__(other)
+        self.changed = True
+        return self
+
+    def update(self, other: dict) -> None:
+        """
+        Merges the contents of a dict into the session.
+        """
+        self |= other
+
```
Comment (ArneGudermann) [inline src/viur/core/session.py:190]:
```suggestion
        self |= other
        self.changed = True
```
We change the session data here to.

**Review by ArneGudermann — CHANGES_REQUESTED** [review]

File: src/viur/core/session.py (hunk 3 above)
Comment (phorward) [reply src/viur/core/session.py:189]: No, because `|=`-operator triggers __ior__, which sets the changed flag.

**Comment by phorward** [comment]
> What about `Session.pop()` and where is `Session.setdefault()` from #1140?

@KadirBalku `Session.pop()` must be implemented.
@sveneberth `Session.setdefault()` is provided by sub-classing `db.Entity`, which sub-classes `dict`.

File: src/viur/core/session.py (hunk 2 above)
Comment (KadirBalku) [reply src/viur/core/session.py:77]: applied and tested, worked.

File: src/viur/core/session.py [hunk 4]
```diff
@@ -199,20 +161,37 @@ def reset(self) -> None:
...
+    def update(self, other: dict) -> None:
         """
-            Returns all items in the current session.
+        Merges the contents of a dict into the session.
         """
-        return self.session.items()
+        self |= other
 
-    def setdefault(self, key: str, value: t.Any) -> t.Any:
+    def pop(self, key: str) -> None:
```
Comment (sveneberth) [inline src/viur/core/session.py:189]: This is not how `pop()` of a container usually works. Take a look at [`dict.pop()` in python docs](https://docs.python.org/3/library/stdtypes.html#dict.pop). It must return the pop'ed element or the default value (if provided)

**Review by sveneberth — CHANGES_REQUESTED** [review]

File: src/viur/core/session.py (hunk 4 above)
Comment (KadirBalku) [reply src/viur/core/session.py]: I have updated #1153

File: src/viur/core/session.py [hunk 5]
```diff
@@ -199,20 +161,48 @@ def reset(self) -> None:
...
+        default = kwargs.get("default", None)
+
+        if default is None:
+            ret = super().pop(key)
+        else:
+            ret = super().pop(key, default)
+
+        self.changed = True
+
+        return ret
```
Comment (sveneberth) [inline src/viur/core/session.py:205]: In your implementation `default` cannot be provided as positional argument and you allow more argument than necessary.
I would work in this case with an sentinel.


```py
_SENTINEL : t.Final[object] = object()
```


```suggestion
    def pop(self, key: str, default=_SENTINEL) -> t.Any:
        """
        Delete a specified key from the session.
        
        If key is in the session, remove it and return its value, else return default. 
        If default is not given and key is not in the session, a KeyError is raised.
        """
        if key in self or default is _SENTINEL:
            value = super().pop(key)
            self.changed = True
            return value
        return default
```

The `KeyError` will be thrown through the super call

**Review by sveneberth — CHANGES_REQUESTED** [review]

**Comment by phorward** [comment]
@KadirBalku we should discuss about dropping Python 3.10 support with viur-core>=3.7

File: src/viur/core/session.py [hunk 6]
```diff
@@ -199,20 +163,46 @@ def reset(self) -> None:
...
+        self |= other
+
+    def pop(self, key: str, default=_SENTINEL) -> t.Any:
+        """
+        Delete a specified key from the session.
+
+        If key is in the session, remove it and return its value, else return default.
+        If default is not given and key is not in the session, a KeyError is raised.
+        """
+        print(f"{_SENTINEL=}")
```
Comment (sveneberth) [inline src/viur/core/session.py:198]:
```suggestion
```

**Review by phorward — APPROVED** [review]

**Review by sveneberth — APPROVED** [review]
LGTM, But we should discuss the python 3.10 issue first, before we merge this. So let's wait until Friday.

**Review by phorward — APPROVED** [review]

**Review by sveneberth — APPROVED** [review]

**Review by ArneGudermann — APPROVED** [review]

**Pull request merged**